lib/vssh/libssh.c

```diff
@@ -81,18 +81,33 @@
 #include "multiif.h"
 #include "select.h"
 #include "warnless.h"
+#include "curl_path.h"
 
 /* for permission and open flags */
+#ifdef HAVE_SYS_TYPES_H
 #include <sys/types.h>
+#endif
+#ifdef HAVE_SYS_STAT_H
 #include <sys/stat.h>
+#endif
+#ifdef HAVE_UNISTD_H
 #include <unistd.h>
+#endif
+#ifdef HAVE_FCNTL_H
 #include <fcntl.h>
+#endif
+
+#ifndef S_IFMT
+#define S_IFMT  00170000
+#endif
+#ifndef S_IFLNK
+#define S_IFLNK  0120000
+#endif
```

Comment:
I don't think these should be defined, they may have different values on different platforms. For example my CRT uses 0xF000 for IFMT mask and there is no IFLNK file type. Also the code these are used for doesn't make sense to me:

~~~c
          if((sshc->readdir_attrs->flags & SSH_FILEXFER_ATTR_PERMISSIONS) &&
             ((sshc->readdir_attrs->permissions & S_IFMT) ==
              S_IFLNK)) {
~~~
So these are remote permissions? What if the client has IFMT, IFLNK values different from the server?

/cc @nmav

Reply:
Ah, I found these in liibssh/sftp.h:
~~~
#define SSH_S_IFMT   00170000
#define SSH_S_IFLNK  0120000